src/code/z_eventmgr.c

```diff
@@ -1,52 +1,549 @@
+    return hudVisibility;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_GetCutscene.s")
+ActorCutscene* ActorCutscene_GetCutsceneImpl(s16 csId) {
+    if (csId < CS_ID_GLOBAL_78) {
+        return &sActorCutsceneList[csId];
+    } else {
+        csId -= CS_ID_GLOBAL_78;
+        return &sGlobalActorCutscenes[csId];
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_GetAdditionalCutscene.s")
+void ActorCutscene_Init(PlayState* play, ActorCutscene* actorCutsceneList, s16 num) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_GetLength.s")
+    sActorCutsceneList = actorCutsceneList;
+    sActorCutsceneCount = num;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F2138.s")
+    for (i = 0; i < ARRAY_COUNT(sWaitingActorCutsceneList); i++) {
+        sWaitingActorCutsceneList[i] = 0;
+        sNextActorCutsceneList[i] = 0;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F2178.s")
+    sActorCsMgr.endCsId = CS_ID_NONE;
+    sActorCsMgr.play = play;
+    sActorCsMgr.length = -1;
+    sActorCsMgr.targetActor = NULL;
+    sActorCsMgr.subCamId = SUB_CAM_ID_DONE;
+    sActorCsMgr.isCameraStored = false;
+    sActorCsMgr.csId = sActorCsMgr.endCsId;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_GetCurrentSubCamId.s")
+/**
+ * Store camera into subCam 2, and keep subCam 2 INACTIVE to preserve the struct.
+ */
+void ActorCutscene_StoreCamera(Camera* camera) {
+    if (camera != NULL) {
+        memcpy(&sActorCsMgr.play->subCameras[2], camera, sizeof(Camera));
+        sActorCsMgr.play->subCameras[2].camId = camera->camId;
+        Camera_ChangeStatus(&sActorCsMgr.play->subCameras[2], CAM_STATUS_INACTIVE);
+        sActorCsMgr.isCameraStored = true;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F21CC.s")
+void ActorCutscene_ClearWaiting(void) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F22C4.s")
+    for (i = 0; i < ARRAY_COUNT(sWaitingActorCutsceneList); i++) {
+        sWaitingActorCutsceneList[i] = 0;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_SetReturnCamera.s")
+void ActorCutscene_ClearNextCutscenes(void) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sNextActorCutsceneList); i++) {
+        sNextActorCutsceneList[i] = 0;
+    }
+}
+
+s16 ActorCutscene_MarkNextCutscenes(void) {
+    s16 bit;
+    s32 i;
+    s32 j;
+    s32 count = 0;
+    s16 actorCsIdMax = CS_ID_NONE;
+    s16 priorityMax = SHT_MAX;
+    s16 csId;
+    s16 priority;
+
+    for (i = 0; i < ARRAY_COUNT(sNextActorCutsceneList); i++) {
+        for (bit = 1, j = 0; j < 8; j++) {
+            if (sWaitingActorCutsceneList[i] & bit) {
+                csId = (i << 3) | j;
+                priority = ActorCutscene_GetCutsceneImpl(csId)->priority;
+
+                if ((priority ^ 0) == -1) {
```

Comment:
Priority -1 means guaranteed to play?

Reply:
Yeah, it adds it to `sNextActorCutsceneList`, but the issue I'm seeing is that it doesn't prevent a second entry from being added to `sNextActorCutsceneList`?
i.e.
```c
if (actorCsIdMax != CS_ID_NONE) {
    sNextActorCutsceneList[actorCsIdMax >> 3] |= 1 << (actorCsIdMax & 7);
}
```
still runs. And also, if multiple waiting cutscenes all have priority `-1`, does that mean they all get added?

Reply:
Yeah, so `-1` priority just means `ActorCutscene_GetCanPlayNext` will always return true. Nothing from `eventMgr` will trigger the cutscene to start, that's still reliant on the external actor/system to start the cutscene.